model/cluster.go

```diff
@@ -105,3 +112,63 @@ var clusterVersionMatcher = regexp.MustCompile(`^(([0-9]{1,3}.[0-9]{1,3}.[0-9]{1
 func ValidClusterVersion(name string) bool {
 	return clusterVersionMatcher.MatchString(name)
 }
+
+// UpdateUtilityMetadata takes a map of string to string representing
+// any metadata related to the utility group and stores it as a []byte
+// in Cluster so that it can be inserted into the database
+func (c *Cluster) UpdateUtilityMetadata(versions map[string]string) error {
+	var metadata []byte
+	metadata, err := json.Marshal(versions)
+	if err != nil {
+		return errors.Wrapf(err, "failed to marshal provided utility metadata map %#v", versions)
+	}
+
+	c.UtilityMetadata = metadata
+	return nil
+}
+
+type utilityVersions struct {
+	Prometheus string
+	Nginx      string
+	Fluentbit  string
+}
+
+// GetUtilityVersion fetches the desired version of a utility from the
+// Cluster object
+func (c *Cluster) GetUtilityVersion(utility string) (string, error) {
```

Comment:
As you probably know, the caching-on-cluster-object in this function is a departure from how we simply return metadata objects for things like the provisioner metadata.

Did you find that simply returning an object here led to needing to unmarshal the same object multiple times for each supervisor flow? Considering that each time the supervisors wake up to work they will be re-querying the database for cluster objects, the caching done here should be short-lived and will need to be redone anyway.

I am mainly curious if this may just be extra complexity for now. If we do want to cache metadata, it may be best handled in a PR that converts all of these values to that behavior.

Thoughts?

Reply:
augh, DMing you, I apparently pushed changes I didn't mean to while resolving conflicts from your PR @gabrieljackson

Reply:
I just didn't want to do json deserialization/serialization every time someone requested this object. To be honest, I don't like storing the data as a []byte but our data models are a leaky abstraction, reflecting the underlying SQL data type. So if it's a BYTEA in the database, it has to be a []byte in Cluster, which means that since that []byte is actually JSON, we can either serialize/deserialize every single time, or we can implement an intermediate Cluster representation in memory after taking it out of the database, or we can revert back to my old schema with typed fields for each piece of metadata that we want to store, or we can cache the native representation of the deserialized data like this and use a Getter, which is the simplest solution to avoid (de)serializing the JSON every time we want to update or read from this field.